README.md

```diff
@@ -671,6 +672,67 @@ of hook, see the [git-hooks documentation][GHD].
 
 [GHD]: https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks
 
+### Adding Existing Line-Aware Commands
+
+Or in other words "low-code error format support."
+
+If you use tools that analyze files and report their findings line-by-line,
+and that Overcommit does not yet support, you may be able to integrate them
+with Overcommit without writing any Ruby code in a similar way as
+[for existing Git hooks](#adding-existing-git-hooks).
+
+These special line-aware command hooks behave and are configured the same way
+as the Git ones, except only file arguments get passed to them.
+Also they must have the `ad_hoc` option, so that, using the command output:
+- differentiating between warnings and errors becomes possible
+- modified lines can be detected and acted upon as defined by
+  the `problem_on_unmodified_line`, `requires_files`, `include` and `exclude`
+  [hook options](#hook-options)
+
+**Warning**: Only the command's standard output stream is considered for now,
+*not* its standard error stream.
+
+To differentiate between warning and error messages,
+the `warning_message_type_pattern` suboption may be specified:
+the `type` field of the `message_pattern` regexp below must then include
+the `warning_message_type_pattern` option's text.
+
+The `message_pattern` suboption specifies the format of the command's messages.
+It is a optional [(Ruby) regexp][RubyRE], which if present must at least define
+a `file` [named capture group][RubyRENCG].
+The only other allowed ones are `line` and `type`, which when specified
+enable detection of modified lines and warnings respectively.
+
+**Note**: The default value for this option is often adequate:
+it generalizes the quasi-standard [GNU/Emacs-style error format][GNUEerrf],
+adding the most frequently used warning syntax to it.
+
+For example:
+
+```yaml
+PreCommit:
+  CustomScript:
+    enabled: true
+    command: './bin/custom-script'
+    ad_hoc:
```

Comment:
Rather than having this high-level key, let's flatten this list so that if `command` is detected to be a script one can optionally specify `message_pattern` and other options to tweak how the output of that script is processed.

This would fit with the path forward discussed earlier, where built-in hooks change their implementations to use the `message_pattern` option rather than explicitly define their own message extraction ([example](https://github.com/sds/overcommit/blob/7d94fa55a74ceea705645d565a34ecefaca08ea5/lib/overcommit/hook/pre_commit/rubo_cop.rb#L20-L32) would simply be transformed into the invocation of a command and filter the output using the defined `message_pattern`, though perhaps RuboCop specifically would still need a custom one. Hopefully the idea is clear).

Reply:
IIUC to sum it up you'd like a flat YAML structure where the new `message_pattern` option's presence activates message extraction. If that's fine with you could we agree on the other details first? I would be glad if I could avoid too many rounds of PR rework.

Reply:
You got it—a flatter YAML structure would be slightly easier to remember, and doesn't require the introduction of the word "ad hoc" to the configuration.

Reply:
OK I now started using the term "line-aware" more systemically internally too for my new type of ad-hoc hooks, and "ad-hoc" is nowhere to be seen anymore in the configuration and docs.